04-onion-routing.md

```diff
@@ -593,75 +601,78 @@ Where `hmac` is an HMAC authenticating the remainder of the packet, with a key
+    - SHOULD use constant `ammag` and `um` keys to de-obfuscate the route length.
 
 The association between the forward and return packets is handled outside of
-this protocol, e.g. via association with an HTLC in a payment channel.
+this onion routing protocol, e.g. via association with an HTLC in a payment
+channel.
 
 ## Failure Messages
 
 The failure message encapsulated in `failuremsg` has an identical format as
 a normal message: a 2-byte type `failure_code` followed by data applicable
-to that type. Following is a list of `failure_code` values currently supported
-along with their intended use cases.
+to that type. Following is a list of the currently supported `failure_code`
+values and their associated use cases.
 
-Notice that the `failure_code`s are not of the same type as other message types
-(defined in other BOLTs), as they are not sent directly on the transport layer
+Notice that the `failure_code`s are not of the same type as other message types,
+defined in other BOLTs, as they are not sent directly on the transport layer
 but are instead wrapped inside return packets.
-The numeric values for the `failure_code` may therefore reuse values that are
-also assigned as message types, without any danger of causing collisions.
+The numeric values for the `failure_code` may therefore reuse values, that are
+also assigned to other message types, without any danger of causing collisions.
 
-A node:
+An erring node:
   - MUST select one of these codes when creating an error message.
   - MUST include the appropriate data for that particular error type.
   - if there is more than one error:
-    - SHOULD select the first error listed below.
+    - SHOULD select the first error it encounters from the list below.
 
 The top byte of `failure_code` can be read as a set of flags:
 * 0x8000 (BADONION): unparsable onion encrypted by sending peer
 * 0x4000 (PERM): permanent failure (otherwise transient)
 * 0x2000 (NODE): node failure (otherwise channel)
 * 0x1000 (UPDATE): new channel update enclosed
 
-Any node MAY return one of the following errors:
+Any erring node MAY return one of the following errors:
   - if the `realm` byte is unknown:
     1. type: PERM|1 (`invalid_realm`)
   - if an otherwise unspecified transient error occurs for the entire node:
     1. type: NODE|2 (`temporary_node_failure`)
   - if an otherwise unspecified permanent error occurs for the entire node:
     1. type: PERM|NODE|2 (`permanent_node_failure`)
-  - if a node has requirement advertised in its `node_announcement` `features`
-  which were not present in the onion:
+  - if a node has requirements advertised in its `node_announcement` `features`
+  which were NOT included in the onion:
     1. type: PERM|NODE|3 (`required_node_feature_missing`)
 
-A _forwarding node_ MAY, but the _final node_ MUST NOT, return one of the following
+A _return-forwarding node_ MAY, but a _final node_ MUST NOT, return one of the following
```

Comment:
_forwarding node_ was right here: it's forwarding the packet, when it encounters an error...

Reply:
fixed